lms/djangoapps/shoppingcart/models.py

```diff
@@ -1235,6 +1238,30 @@ class CouponRedemption(models.Model):
     coupon = models.ForeignKey(Coupon, db_index=True)
 
     @classmethod
+    def remove_code_redemption(cls, item, user):
+        """
+        If an item removed from shopping cart then we will remove
+        the corresponding redemption info of coupon code
+        """
+        order_item_course_id = getattr(item, 'course_id')
+        try:
+            # Try to remove redemption information of coupon code, If exist.
+            coupon_redemption = cls.objects.get(
+                user=user,
+                coupon__course_id=order_item_course_id if order_item_course_id else CourseKeyField.Empty,
+                order=item.order_id
+            )
+            coupon_redemption.delete()
```

Comment:
The existing method `delete_coupon_redemption()` deletes all the redemption entries against the user cart. Consider a scenario where the user has 2 items in the cart and the same coupon code is being used by both the items. If one of the item get's expired, then it will delete all the coupon redemption entries.